larpandoracontent/LArThreeDReco/LArCosmicRay/ThreeViewDeltaRayMatchingAlgorithm.cc

```diff
@@ -0,0 +1,164 @@
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeViewDeltaRayMatchingAlgorithm::CalculateOverlapResult(const Cluster *const pClusterU, const Cluster *const pClusterV, const Cluster *const pClusterW,
+    DeltaRayOverlapResult &overlapResult) const
+{
+    float chiSquaredSum(0.f);
+    unsigned int nSamplingPoints(0), nMatchedSamplingPoints(0);
+    XOverlap xOverlapObject(0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f);
+    
+    StatusCode statusCode(this->PerformThreeViewMatching(pClusterU, pClusterV, pClusterW, chiSquaredSum, nSamplingPoints, nMatchedSamplingPoints, xOverlapObject));
+
+    if (statusCode != STATUS_CODE_SUCCESS)
+        return statusCode;
+
+    PfoList commonMuonPfoList;
+    this->FindCommonMuonParents(pClusterU, pClusterV, pClusterW, commonMuonPfoList);
+
+    if (commonMuonPfoList.empty())
+        return STATUS_CODE_NOT_FOUND;
+
+    overlapResult = DeltaRayOverlapResult(nMatchedSamplingPoints, nSamplingPoints, chiSquaredSum, xOverlapObject, commonMuonPfoList);
+    
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeViewDeltaRayMatchingAlgorithm::FindCommonMuonParents(const Cluster *const pClusterU, const Cluster *const pClusterV, const Cluster *const pClusterW,
+    PfoList &commonMuonPfoList) const
+{
+    ClusterList consideredClustersU, consideredClustersV, consideredClustersW;
+    PfoList nearbyMuonPfosU, nearbyMuonPfosV, nearbyMuonPfosW;
+    
+    this->GetNearbyMuonPfos(pClusterU, consideredClustersU, nearbyMuonPfosU);
+
+    if (nearbyMuonPfosU.empty())
+        return;
+    
+    this->GetNearbyMuonPfos(pClusterV, consideredClustersV, nearbyMuonPfosV);
+
+    if (nearbyMuonPfosV.empty())
+        return;
+    
+    this->GetNearbyMuonPfos(pClusterW, consideredClustersW, nearbyMuonPfosW);
+
+    if (nearbyMuonPfosW.empty())
+        return;    
+
+    for (const ParticleFlowObject *const pNearbyMuonU : nearbyMuonPfosU)
+    {
+        for (const ParticleFlowObject *const pNearbyMuonV : nearbyMuonPfosV)
+        {
+            for (const ParticleFlowObject *const pNearbyMuonW : nearbyMuonPfosW)
+            {
+                if ((pNearbyMuonU == pNearbyMuonV) && (pNearbyMuonV == pNearbyMuonW))
+                    commonMuonPfoList.push_back(pNearbyMuonU);
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeViewDeltaRayMatchingAlgorithm::ExamineOverlapContainer()
+{
+    unsigned int repeatCounter(0);
+
+    for (TensorToolVector::const_iterator toolIter = m_algorithmToolVector.begin(); toolIter != m_algorithmToolVector.end(); )
```

Comment:
Can we replace this with:

```
for (auto tool : m_algorithmToolVector)
    for (unsigned int rep = 0; rep <= m_nMaxTensorToolRepeats; ++rep)
        if (!tool->Run(this, this->GetMatchingControl().GetOverlapTensor()))
            break;
```

? (Check that I got the loop counter right - and open to thoughts on if and where we want braces here - they're not necessary, but perhaps desirable for clarity?)

Reply:
i think i'm inclined to go with 

for (auto tool : m_algorithmToolVector)
{
    for (unsigned int rep = 0; rep <= m_nMaxTensorToolRepeats; ++rep)
    {
        if (!tool->Run(this, this->GetMatchingControl().GetOverlapTensor()))
            break;
    }
}

?

Reply:
obvioulsy with the right indentations...